File: src/main/java/ru/gdlbo/search/searcher/controllers/FileHistoryController.java
package ru.gdlbo.search.searcher.controllers;

import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import ru.gdlbo.search.searcher.repository.FileHistory;
import ru.gdlbo.search.searcher.repository.FileInfo;

import java.io.File;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

@Controller
public class FileHistoryController {
    // This method is responsible for retrieving the history of a file
    @GetMapping("/api/history")
    public ResponseEntity<List<FileHistory>> getFileHistory(@RequestParam String fileName) {
        System.out.println("Request received to get history for file: " + fileName);

        File file = new File(fileName);
        File parentDir = file.getParentFile();
        File hiddenDir = new File(parentDir, ".history");

        // Get all files in the hidden directory with the given file name
        File[] files = hiddenDir.listFiles((dir, name) -> {
            String baseName = file.getName();
            int lastDotIndex = baseName.lastIndexOf('.');
            if (lastDotIndex != -1) {
                baseName = baseName.substring(0, lastDotIndex);
            }

            String currentFileNameWithoutExtension = name;
            int currentFileLastDotIndex = currentFileNameWithoutExtension.lastIndexOf('.');
            if (currentFileLastDotIndex != -1) {
                currentFileNameWithoutExtension = currentFileNameWithoutExtension.substring(0, currentFileLastDotIndex);
            }

            return currentFileNameWithoutExtension.startsWith(baseName);
        });

        // If there are no files, return an empty list
        if (files == null) {
            System.out.println("No history found for file: " + fileName);
            files = new File[0];
        }

        // Sort the files by timestamp
        List<File> fileList = new ArrayList<>(Arrays.asList(files));
        fileList.sort(Comparator.comparingLong(File::lastModified));
        String location = file.getAbsolutePath();

        List<FileHistory> fileHistory = fileList.stream()
                .map(f -> new FileHistory(extractDateFromFileName(f), location))
                .toList();

        System.out.println("Returning history for file: " + fileName);

        return ResponseEntity.ok(fileHistory);
    }

    private static String extractDateFromFileName(File file) {
        String fileName = file.getName();
        String datePattern = "\\d{4}-\\d{2}-\\d{2}_\\d{2}-\\d{2}-\\d{2}";

        Pattern pattern = Pattern.compile(datePattern);
        Matcher matcher = pattern.matcher(fileName);

        if (matcher.find()) {
            return matcher.group(0).replace("_", " ");
        } else {
            return "N/A";
        }
    }

    // This method is responsible for removing a file from the history
    @GetMapping("/api/removeFromHistory")
    public String removeFileFromHistory(@RequestParam String filePath, Authentication authentication) {
        System.out.println("Request received to remove file: " + filePath);

        File fileToRemove = new File(filePath);

        // Check to not abuse deletion of files
        if (!fileToRemove.getAbsolutePath().contains(".history") && !authentication.getAuthorities().contains(new SimpleGrantedAuthority("ROLE_ADMIN"))) {
            System.out.println("Attempt to delete file outside of history directory: " + filePath);
            return "redirect:/error";
        }

        // Remove the file from the hidden directory
        if (!fileToRemove.delete()) {
            System.out.println("Failed to delete file: " + filePath);
            return "redirect:/error";
        }

        System.out.println("Deleted file: " + filePath);

        return "redirect:/search";
    }
}

File: src/main/java/ru/gdlbo/search/searcher/controllers/api/FileHistoryController.kt
package ru.gdlbo.search.searcher.controllers.api

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.core.Authentication
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestParam
import ru.gdlbo.search.searcher.repository.files.FileHistory
import java.io.File
import java.util.regex.Pattern

@Controller
class FileHistoryController {
    // Этот метод отвечает за получение истории изменений файла
    @GetMapping("/api/history")
    fun getFileHistory(@RequestParam fileName: String): ResponseEntity<List<FileHistory>> {
        println("Запрос на получение истории файла: $fileName")

        val file = File(fileName) // Создать объект File для исходного файла
        val files = getFiles(fileName) // Получить список файлов-историй

        // Если список файлов пуст, вернуть пустой список
        if (files.isNullOrEmpty()) {
            println("История для файла $fileName не найдена")
            return ResponseEntity.ok(emptyList())
        }

        // Преобразовать список файлов в List и отсортировать по дате изменения
        val fileList: List<File> = files.toList().sortedBy { it.lastModified() }

        val location = file.absolutePath // Получить абсолютный путь исходного файла

        val fileHistory = fileList.map { f -> FileHistory(extractDateFromFileName(f), location) } // Создать объекты FileHistory

        println("Возвращаем историю файла: $fileName")

        return ResponseEntity.ok(fileHistory)
    }

    @GetMapping("/api/removeAllFromHistory")
    fun deleteFileHistory(@RequestParam fileName: String): ResponseEntity<String> {
        println("Запрос на удаление всей истории файла: $fileName")

        val files = getFiles(fileName) // Получить список файлов-историй

        if (files != null) {
            for (historyFile in files) {
                if (historyFile.delete()) {
                    println("Удален файл истории: " + historyFile.name)
                } else {
                    println("Не удалось удалить файл истории: " + historyFile.name)
                    return ResponseEntity("Не удалось удалить файл истории", HttpStatus.BAD_REQUEST)
                }
            }
        } else {
            println("История для файла $fileName не найдена")
            return ResponseEntity("История для файла не найдена", HttpStatus.BAD_REQUEST)
        }
        return ResponseEntity.ok("Файл истории удален")
    }

    // Этот метод отвечает за удаление отдельного файла из истории
    @GetMapping("/api/removeFromHistory")
    fun removeFileFromHistory(@RequestParam filePath: String, authentication: Authentication): ResponseEntity<String> {
        println("Запрос на удаление файла истории: $filePath")

        val fileToRemove = File(filePath) // Создать объект File для файла на удаление

        // Проверка на попытку удаления файлов вне директории истории
        if (!fileToRemove.absolutePath.contains(".history") && !authentication.authorities.contains(
                SimpleGrantedAuthority("ROLE_ADMIN")
            )
        ) {
            println("Попытка удалить файл вне директории истории: $filePath")
            return ResponseEntity.ok("Не удалось удалить файл")
        }

        // Удалить файл из скрытой директории
        if (!fileToRemove.delete()) {
            println("Не удалось удалить файл: $filePath")
            return ResponseEntity.ok("Не удалось удалить файл")
        }

        println("Файл удален: $filePath")

        return ResponseEntity.ok("Файл удален")
    }

    companion object {
        private fun getFiles(path: String): Array<File>? {
            val file = File(path)
            val parentDir = file.parentFile
            val hiddenDir = File(parentDir, ".history")

            return hiddenDir.listFiles { _: File?, name: String ->
                var baseName = file.name
                val lastDotIndex = baseName.lastIndexOf('.')
                if (lastDotIndex != -1) {
                    baseName = baseName.substring(0, lastDotIndex)
                }

                var currentFileNameWithoutExtension = name
                val currentFileLastDotIndex = currentFileNameWithoutExtension.lastIndexOf('.')
                if (currentFileLastDotIndex != -1) {
                    currentFileNameWithoutExtension =
                        currentFileNameWithoutExtension.substring(0, currentFileLastDotIndex)
                }
                currentFileNameWithoutExtension.startsWith(baseName)
            }
        }

        private fun extractDateFromFileName(file: File): String {
            val fileName = file.name
            val datePattern = "\\d{4}-\\d{2}-\\d{2}_\\d{2}-\\d{2}"

            val pattern = Pattern.compile(datePattern)
            val matcher = pattern.matcher(fileName)

            return if (matcher.find()) {
                matcher.group(0).replace("_", " ")
            } else {
                "N/A"
            }
        }
    }
}